File: 2020/21_AllergenAssessment/test_label.py
# ======================================================================
# Allergen Assessment
#   Advent of Code 2020 Day 21 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                       t e s t _ l a b e l . p y
# ======================================================================
"Test Label for Advent of Code 2020 day 21, Allergen Assessment"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import label

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE_TEXT = "mxmxvkd kfcds sqjhc nhms (contains dairy, fish)"

# ======================================================================
#                                                              TestLabel
# ======================================================================


class TestLabel(unittest.TestCase):  # pylint: disable=R0904
    "Test Label object"

    def test_empty_init(self):
        "Test the default Label creation"

        # 1. Create default Label object
        myobj = label.Label()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, None)
        self.assertEqual(len(myobj.ingredients), 0)
        self.assertEqual(len(myobj.allergens), 0)

    def test_text_init(self):
        "Test the Label object creation from text"

        # 1. Create Label object from text
        myobj = label.Label(text=EXAMPLE_TEXT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 47)
        self.assertEqual(len(myobj.ingredients), 4)
        self.assertEqual(len(myobj.allergens), 2)

        # 3. Check methods
        self.assertEqual(myobj.has_ingredient('mxmxvkd'), True)
        self.assertEqual(myobj.has_ingredient('kfcds'), True)
        self.assertEqual(myobj.has_ingredient('dairy'), False)
        self.assertEqual(myobj.has_allergen('dairy'), True)
        self.assertEqual(myobj.has_allergen('fish'), True)
        self.assertEqual(myobj.has_allergen('soy'), False)
        self.assertEqual(myobj.has_allergen('kfcds'), False)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                     t e s t _ l a b e l . p y                  end
# ======================================================================

File: 2020/21_AllergenAssessment/test_label.lua
-- ======================================================================
-- Allergen Assessment
--   Advent of Code 2020 Day 21 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ l a b e l . l u a
-- ======================================================================
-- Test Label for Advent of Code 2020 day 21, Allergen Assessment

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local label = require('label')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = "mxmxvkd kfcds sqjhc nhms (contains dairy, fish)"

-- ======================================================================
--                                                              TestLabel
-- ======================================================================

function test_empty_init()
  -- Test the default Label creation

  -- 1. Create default Label object
  local myobj = label:Label()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(#myobj.ingredients, 0)
  luaunit.assertEquals(#myobj.allergens, 0)

end

function test_text_init()
  -- Test the Label object creation from text

  -- 1. Create Shopping object from text
  local myobj = label:Label({text=EXAMPLE_TEXT})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.text, EXAMPLE_TEXT)
  luaunit.assertEquals(#myobj.ingredients, 4)
  luaunit.assertEquals(#myobj.allergens, 2)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:has_ingredient('mxmxvkd'), true)
  luaunit.assertEquals(myobj:has_ingredient('kfcds'), true)
  luaunit.assertEquals(myobj:has_ingredient('dairy'), false)
  luaunit.assertEquals(myobj:has_allergen('dairy'), true)
  luaunit.assertEquals(myobj:has_allergen('fish'), true)
  luaunit.assertEquals(myobj:has_allergen('soy'), false)
  luaunit.assertEquals(myobj:has_allergen('kfcds'), false)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                    t e s t _ l a b e l . p y                   end
-- ======================================================================
